generate/exportedtf/exportedtf.go

```diff
@@ -0,0 +1,216 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exportedtf
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/madlambda/spells/errutil"
+	"github.com/mineiros-io/terramate"
+	"github.com/mineiros-io/terramate/config"
+	"github.com/mineiros-io/terramate/hcl"
+	"github.com/mineiros-io/terramate/hcl/eval"
+	"github.com/mineiros-io/terramate/stack"
+	"github.com/rs/zerolog/log"
+)
+
+// StackTf represents all exported terraform code for a stack,
+// mapping the exported code name to the actual Terraform code.
+type StackTf struct {
+	tfcode map[string]Body
+}
+
+// Body represents exported Terraform code from a single block.
+// Is contains parsed and evaluated code on it.
+type Body struct {
+	body []byte
+}
+
+const (
+	ErrInvalidBlock errutil.Error = "invalid export_as_terraform block"
+	ErrEval         errutil.Error = "evaluating export_as_terraform block"
+)
+
+// ExportedCode returns all exported code, mapping the name to its
+// equivalent generated code.
+func (s StackTf) ExportedCode() map[string]Body {
+	cp := map[string]Body{}
+	for k, v := range s.tfcode {
+		cp[k] = v
+	}
+	return cp
+}
+
+// String returns a string representation of the Terraform code
+// or an empty string if the config itself is empty.
+func (b Body) String() string {
+	return string(b.body)
+}
+
+// Load loads from the file system all export_as_terraform for
+// a given stack. It will navigate the file system from the stack dir until
+// it reaches rootdir, loading export_as_terraform and merging them appropriately.
+//
+// More specific definitions (closer or at the stack) have precedence over
+// less specific ones (closer or at the root dir).
+//
+// Metadata and globals for the stack are used on the evaluation of the
+// export_as_terramate blocks.
+//
+// The returned result only contains evaluated values.
+//
+// The rootdir MUST be an absolute path.
+func Load(rootdir string, sm stack.Metadata, globals *terramate.Globals) (StackTf, error) {
+	stackpath := filepath.Join(rootdir, sm.Path)
+	logger := log.With().
+		Str("action", "exportedtf.Load()").
+		Str("path", stackpath).
+		Logger()
+
+	logger.Trace().Msg("loading export_as_terraform blocks.")
+
+	exportBlocks, err := loadExportBlocks(rootdir, stackpath)
+	if err != nil {
+		return StackTf{}, fmt.Errorf("loading exported terraform code: %w", err)
+	}
+
+	evalctx, err := newEvalCtx(stackpath, sm, globals)
+	if err != nil {
+		return StackTf{}, fmt.Errorf("preparing to eval exported terraform code: %v", err)
```

Comment:
why are we using `%v` instead of `%w` as in line 91?

Reply:
You use %w for wrapping when you intend to expose the error for further inspection and use %v or %s if you want the error to be opaque and conceil details (only the error str representation will be added on the new error, but no further type info). More details can be found [here](https://go.dev/blog/go1.13-errors).

Overall there is several different ways to model errors so they can be inspected for details, sentinels are the simplest ones with some tradeoffs on coupling or even leaking details (as mentioned on the blog post). So I try to use wrapping/sentinels only when I want to expose a clear distinction on the nature of a failure. For the creation of the eval ctx I didn't thought it would be useful, but we could add a sentinel if you think it makes sense (like ErrEval).

Reply:
Ah I also add Sentinels usually as I test, since I use then on the tests to check for specific errors. I didn't tested eval ctx creation so far, so that is also a reason it doesn;t have a wrapping. My default is to not wrap/use %v, so I just add the wrap when Im intentionally wanting to expose/test something through the API.